sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -1247,33 +1315,109 @@ public void startBundle(StartBundleContext c) throws IOException {
       KV<ByteString, Iterable<Mutation>> record = c.element();
-      bigtableWriter
-          .writeRecord(record)
-          .whenComplete(
-              (mutationResult, exception) -> {
-                if (exception != null) {
-                  failures.add(new BigtableWriteException(record, exception));
-                }
-              });
+      bigtableWriter.writeRecord(record).whenComplete(handleMutationException(record, window));
       ++recordsWritten;
       seenWindows.compute(window, (key, count) -> (count != null ? count : 0) + 1);
     }
 
+    public BiConsumer<MutateRowResponse, Throwable> handleMutationException(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      return (MutateRowResponse result, Throwable exception) -> {
+        if (exception != null) {
+          if (exception instanceof NotFoundException
+              && !((NotFoundException) exception).isRetryable()) {
+            // This case, of being an NotFoundException and not retryable,
+            // indicates an issue with the data. In general, the likely cause is the user trying to
+            // write to a column family that doesn't exist. The frequency of this is dependent on
+            // how column families are determined upstream of the io. We can't know if this record
+            // failed, or if it was another record in the batch. Thus, we retry each record
+            // individually.
+            retryIndividualRecord(record, window);
+          } else {
+            failures.add(new BigtableWriteException(record, exception));
+          }
+        }
+      };
+    }
+
+    public void retryIndividualRecord(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      try {
+        unbatchedWrites.add(
+            bigtableWriter
+                .writeSingleRecord(record)
+                .whenComplete(
+                    (singleMutationResult, singleException) -> {
+                      if (singleException != null) {
+                        if (singleException instanceof NotFoundException
+                            && !((NotFoundException) singleException).isRetryable()) {
+                          // if we get another NotFoundException, we know this is the bad record.
+                          badRecords.add(
+                              KV.of(new BigtableWriteException(record, singleException), window));
+                        } else {
+                          failures.add(new BigtableWriteException(record, singleException));
+                        }
+                      }
+                    }));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
     @FinishBundle
     public void finishBundle(FinishBundleContext c) throws Exception {
       try {
+
         if (bigtableWriter != null) {
-          bigtableWriter.close();
+          try {
+            bigtableWriter.close();
+          } catch (IOException e) {
```

Comment:
I dont think any of this correct. 
The writer seems to swallow BatchingExceptions?
https://github.com/apache/beam/blob/8bd4a1e525ba5f3256124ae331acf0cc48ad65f9/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java#L522-L525

Reply:
For whatever reason, it looks like the actual exception type is an execution exception, which is unwrapped an passed upwards. I'm not sure why that is the case, but that is the current state.